perma-payments/perma_payments/views.py

```diff
@@ -0,0 +1,428 @@
+import csv
+from datetime import datetime
+from functools import wraps
+import io
+
+from django.conf import settings
+from django.contrib.auth.decorators import login_required
+from django.core.exceptions import ValidationError, ObjectDoesNotExist, MultipleObjectsReturned, PermissionDenied
+from django.db import transaction
+from django.http import JsonResponse
+from django.shortcuts import render, redirect
+from django.utils.decorators import available_attrs
+from django.views.decorators.debug import sensitive_post_parameters
+from django.views.decorators.http import require_http_methods
+from django.views.decorators.csrf import csrf_exempt
+
+from .constants import *
+from .custom_errors import bad_request
+from .email import send_admin_email
+from .models import *
+from .security import *
+
+import logging
+logger = logging.getLogger(__name__)
+
+#
+# UTILS
+#
+
+FIELDS_REQUIRED_FROM_PERMA = {
+    'subscribe': [
+        'registrar',
+        'amount',
+        'recurring_amount',
+        'recurring_frequency',
+        'recurring_start_date'
+    ],
+    'update': [
+        'registrar'
+    ],
+    'subscription': [
+        'registrar'
+    ],
+    'cancel_request': [
+        'registrar'
+    ]
+}
+
+FIELDS_REQUIRED_FOR_CYBERSOURCE = {
+    'subscribe': [
+        'access_key',
+        'amount',
+        'currency',
+        'locale',
+        'payment_method',
+        'profile_id',
+        'recurring_amount',
+        'recurring_frequency',
+        'recurring_start_date',
+        'reference_number',
+        'signed_date_time',
+        'transaction_type',
+        'transaction_uuid'
+    ],
+    'update': [
+        'access_key',
+        'allow_payment_token_update',
+        'locale',
+        'payment_method',
+        'payment_token',
+        'profile_id',
+        'reference_number',
+        'signed_date_time',
+        'transaction_type',
+        'transaction_uuid'
+    ]
+}
+
+FIELDS_REQUIRED_FROM_CYBERSOURCE = {
+    'cybersource_callback': [
+        'req_transaction_uuid',
+        'decision',
+        'reason_code',
+        'message'
+    ]
+}
+
+SENSITIVE_POST_PARAMETERS = [
+    'payment_token',
+    'req_access_key',
+    'req_bill_to_address_city',
+    'req_bill_to_address_country',
+    'req_bill_to_address_line1',
+    'req_bill_to_address_postal_code',
+    'req_bill_to_address_state',
+    'req_bill_to_email',
+    'req_bill_to_forename',
+    'req_bill_to_surname',
+    'req_card_expiry_date',
+    'req_card_number',
+    'req_payment_token',
+    'req_profile_id',
+    'signature'
+]
+
+
+def redact(post):
+    return {k: v for (k, v) in post.items() if k not in SENSITIVE_POST_PARAMETERS}
```

Comment:
is this used anywhere? Maybe just move it into testing?

Reply:
Ha!! What a funny bug you found! I'm supposed to be calling it in 325 (https://github.com/anastasia/perma-payments/pull/1/files#diff-b36ac4c36c265a1586de521a797dfb9cR325). I was refactoring, and was like, "This logic should be pulled out into a helper function for clarity and testing!" So I made the helper function, and tested it, and the helper function logic is identical to the original logic, so the tests pass....... but I'm never actually calling it. Great catch!

The solution is to call it where I intended to: it's for redacting sensitive information before it hits the logs. Not a testing function.

Reply:
Do you have any concerns of having sensitive data stored in memory at all? I *think* that might be a vulnerability, but maybe this is something we ask @jcushman?

Reply:
So unfortunately there's nothing I can do to avoid it. It isn't TRULY sensitive data (e.g., not credit card numbers), just stuff we shouldn't be logging, like email addresses and addresses, and there's no way to prevent CyberSource from sending it to use. By use of Django's sensitive params and sensitive variables decorators, and by use of this redact function, I'm pretty sure we're doing the best we can. Though I should definitely take another pass and make sure I used sensitive params / variables everywhere necessary.